### ISSUE TITLE
Add location to errors

### ISSUE THREAD

**Issue by jpramosi** [mcmah309/error_set#4]
Hi, a really nice concept to handle and group errors.
I would like to suggest to also include information about the location of the error with [std::panic::location](https://doc.rust-lang.org/core/panic/struct.Location.html#examples-1). It could be useful in large systems to trace it back.
Together with this [feature](https://github.com/mcmah309/error_set/issues/3#issuecomment-2248635159) it would be my _dream crate_ for error handling.

**Comment by mcmah309** [comment]
How would you suggest including the location information- what would you like to see as a user? Probably something that should have a feature gate.

**Comment by jpramosi** [comment]
To be honest, I have settled with my own thiserror fork with a few experimental features for now.

While I have tinkered between different error libraries I have found a neat way to include location- and other data.
I used a generic struct for errors and an user specified enum for the types of errors as template type.
Then putting this type in a blanket struct in the user code to implement any kinds of traits (the code would be generated by the proc-macro). The location, name and other data is stored in the struct itself. With the experimental features (try_trait_v2) I could trace every error and put this in a RefCell vector. In the end the user can retrieve the said data from the generic struct.

I know this is a pretty wild and radical approach for error_set, and I don't expect it to be implemented. But I hope I've given you a rough idea of what I'm doing.

**Comment by mcmah309** [comment]
If you provide some pseudo code of what you are currently doing and/or what you'd like to see from error_set, it would help me get a better idea

**Comment by jpramosi** [comment]
<details>
<summary>Okay, this is a rough sketch of the structure that the error handling uses:</summary>

```rust
#[derive(Clone, Copy, Hash, PartialEq, PartialOrd, Eq, Ord)]
pub struct ErrorId(&'static str);

pub trait ErrorIdentity {
    fn id(&self) -> ErrorId;

    fn origin(&self) -> Option<Arc<dyn std::error::Error>>;
}

/// The items represent an entity where an error was created, propagated or
/// converted from another error in the error chain.
#[derive(Clone, Debug)]
pub struct ErrorItem {
    /// The location of the error.
    pub location: &'static Location<'static>,
    /// The error name.
    pub name: String,
    /// The origin of the error.
    pub origin: Option<Weak<dyn std::error::Error>>,
}

impl ErrorItem {
    #[track_caller]
    pub fn new<T>(error: &T) -> Self
    where
        T: ErrorIdentity,
    {
        let origin = match error.origin() {
            Some(origin) => Some(Arc::downgrade(&origin)),
            None => None,
        };
        ErrorItem {
            location: Location::caller(),
            name: error.id().to_string(),
            origin,
        }
    }
}

/// A shared context which is used in the error propagation chain for tracing purposes.
#[derive(Default, Debug, Clone)]
pub struct TraceContext {
    pub trace: IndexSet<ErrorItem>,
}

impl TraceContext {
    pub fn origin(&self) -> &ErrorItem {
        self.trace
            .first()
            .expect("Context must have at least one or more items.")
    }
}

pub trait ErrorTrace {
    fn insert(&mut self, ctx: ErrorItem);

    fn context_mut(&mut self) -> RefMut<'_, TraceContext>;
}

/// The Span represents a parent which an ErrorItem is referring back to.
/// Multiple of these items can be assigned to a Span.
/// A single Span comprises a single set of error types that are
/// usually defined as an 'Enum' on the user side.
/// In the error propagation chain this Span comprises this set of errors.
#[derive(Debug, Clone)]
pub struct Span<T>
where
    T: std::error::Error + ErrorIdentity,
{
    pub ctx: RefCell<TraceContext>,
    pub created: ErrorItem,
    pub inner: Arc<T>,
}

impl<T> Span<T>
where
    T: std::error::Error + ErrorIdentity + 'static,
{
    #[track_caller]
    pub fn new(ctx: Option<RefCell<TraceContext>>, inner: T) -> Self {
        let inner_owned = Arc::new(inner);
        let inner_ref = Arc::downgrade(&inner_owned);
        let created = ErrorItem {
            location: Location::caller(),
            name: inner_owned.id().to_string(),
            origin: Some(inner_ref),
        };

        dlog!("created: {:#?}", created);
        let ctx = match ctx {
            Some(v) => v,
            None => {
                // With "fn from_residual(r: std::result::Result<Infallible, E>)"
                // the trace can look like it has duplicates, but is actually caused by two different
                // operations in the same function.
                // At the moment it is not required to keep this data around from the particular operation.
                // Instead an IndexSet is used for the trace to only allow unique items.
                let mut trace = IndexSet::with_capacity(TRACE_RESERVE);
                dlog!("insert ({})\n\tin {}", created.name, created.location);
                trace.insert(created.clone());
                RefCell::new(TraceContext { trace })
            }
        };

        Self {
            ctx,
            created,
            inner: inner_owned,
        }
    }
}
```

</details>

<details>
<summary>And this is a part that will be emitted as user-code from the proc-macro (#ty is the error enum):</summary>

```rust
        #[derive(Debug)]
        pub struct ErrorContext(thiserror::Span<#ty>);

        impl ErrorContext {
            #[allow(unused_qualifications)]
            #[track_caller]
            pub fn new(kind: #ty) -> Self {
                return Self(thiserror::Span::new(None, kind));
            }
        }

        #[allow(unused_qualifications)]
        impl std::ops::Deref for ErrorContext {
            type Target = thiserror::Span<#ty>;

            fn deref(&self) -> &Self::Target {
                &self.0
            }
        }

        #[allow(unused_qualifications)]
        impl std::ops::DerefMut for ErrorContext {
            fn deref_mut(&mut self) -> &mut Self::Target {
                &mut self.0
            }
        }

        #[allow(unused_qualifications)]
        impl std::fmt::Display for ErrorContext {
            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                // write!(f, "{}", self.0.inner)
                self.0.inner.fmt(f)
            }
        }

        #[allow(unused_qualifications)]
        impl std::error::Error for ErrorContext {
            fn provide<'a>(&'a self, request: &mut core::error::Request<'a>) {
                request.provide_ref::<thiserror::ErrorItem>(&self.created);
                request.provide_ref::<std::cell::RefCell<thiserror::TraceContext>>(&self.ctx);
            }
        }

        #[allow(unused_qualifications)]
        impl thiserror::ErrorIdentity for ErrorContext {
            fn id(&self) -> thiserror::ErrorId {
                match *self.inner {
                    #(#enum_arms)*
                }
            }

            fn origin(&self) -> Option<std::sync::Arc<dyn std::error::Error>> {
                return Some(self.inner.clone());
            }
        }

        #[allow(unused_qualifications)]
        impl thiserror::ErrorTrace for ErrorContext {
            fn insert(&mut self, ctx: thiserror::ErrorItem) {
                self.ctx.borrow_mut().trace.insert(ctx);
            }

            fn context_mut(&mut self) -> std::cell::RefMut<'_, thiserror::TraceContext> {
                self.ctx.borrow_mut()
            }
        }

        #[allow(unused_qualifications)]
        impl ::core::convert::From<#ty> for ErrorContext {
            #[track_caller]
            fn from(value: #ty) -> Self {
                let ctx = match std::error::request_ref::<std::cell::RefCell<thiserror::TraceContext>>(
                    &value,
                ) {
                    Some(ctx_ref) => Some(ctx_ref.clone()),
                    None => None,
                };
                Self(thiserror::Span::new(ctx, value))
            }
        }

        #(#from_impls)*
```

</details>

<details>
<summary>Usage:</summary>

```rust
#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    Io(#[from] fs::ErrorContext),
    #[error("Invalid parameter: ({0}='{1}')")]
    Parameter(&'static str, String),
    #[error("Could not find {0}")]
    NotFound(PathBuf),
    #[error("Failed to parse {0}")]
    Parse(&'static str),
    #[error(transparent)]
    ParseConvert(#[from] std::num::ParseIntError),
}

#[allow(dead_code)]
#[derive(Clone, Debug)]
pub struct Version {
    pub major: u32,
    pub minor: u32,
    pub patch: u32,
}

fn read_validate(path: PathBuf) -> Result<String, ErrorContext> {
    if !fs::exists(&path)? {
        return err!(Error::NotFound(path));
    }
    Ok(fs::read_to_string(PathBuf::from("invalid-path"))?)
}

fn parse_version(content: &String) -> Result<Version, ErrorContext> {
    for line in content.split("\n") {
        if line.contains("version") {
            let line = line.replace(" ", "");
            let mut data = line.split("version=");
            let mut version = data.next().ok_or(Error::Parse("version"))?.split(".");
            let major = version.next().ok_or(Error::Parse("major"))?;
            let minor = version.next().ok_or(Error::Parse("minor"))?;
            let patch = version.next().ok_or(Error::Parse("patch"))?;
            return Ok(Version {
                major: major.parse()?,
                minor: minor.parse()?,
                patch: patch.parse()?,
            });
        }
    }
    // into() can also be used instead of ErrorContext::new or the err! macro
    Err(Error::Parse("content").into())
}

pub fn get_toml_config(name: &str) -> Result<(String, Version), ErrorContext> {
    if name.is_empty() {
        return err!(Error::Parameter("name", name.into()));
    }

    let content = read_validate(PathBuf::from(name))?;
    let version = parse_version(&content)?;

    return Ok((content, version));
}
```

</details>

The design is not final and this sketch is strongly adapted to the error declaration of thiserror with enums.
But I believe the structure could be probably used here (if desired). But so far I can tell, the users need to enable `#![feature(error_generic_member_access)]` for a working error trace.

> what you'd like to see from error_set

I like the approach of error_set which defines errors like quick_error. What makes this solution complete for me is what another user has already mentioned [here](https://github.com/mcmah309/error_set/issues/3#issuecomment-2248635159).
Basically the error declaration stays the same. What changes and breaks would be the inner workings of error_set plus the user code. That is the reason why I don't mind if this feature will not be implemented here.

Then I couldn't guess what the implications of that feature would be, so I casually suggested it here, haha. I hope you don't mind.

**Comment by mcmah309** [comment]
Would something like [backtrace](https://doc.rust-lang.org/std/backtrace/index.html) suit your needs?

An api like
```rust
error_set! {
    Error = {
        Variant1 {
             /// name: String,
            source: std::io::Error,
            backtrace: std::Backtrace
        },
        Variant2(std::io::Error, std::Backtrace)
    };
}
```

`backtrace`, if not provided by another subset, is captured in the `From` impl (therefore the existence of `backtrace` does not determine sets.

`source` on an inline struct becomes equivalent to a wrapped error, so that also opens the door for inline error structs and wrapped errors to be subsets of each other.

And of course the `std::error::Error` members would be implemented appropriately.

**Comment by jpramosi** [comment]
Yes, it would roughly suit my needs. It is pretty similar how _thiserror_ handles location tracing. But I'm not sure what you mean exactly with source on an inline struct. Is it a blanket struct for the actual error? Anyhow I think the error declaration could be cleaner and more concise if a backtrace field or additional fields are included.
If more error variants are used or additional dynamic data is desired, it may look a bit bloated.

One option could be to wrap the error in a generated struct and "hide" these fields from the user. The struct itself contains the relevant metadata and error type.

If backtrace is no option (performance issues), the location feature could rely on [track_caller](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller). The error record with its location could be inserted into a list or just logged with error_set its tracing feature within the relevant functions or traits on error creation or conversion.

**Comment by mcmah309** [comment]
> But I'm not sure what you mean exactly with source on an inline struct. Is it a blanket struct for the actual error?

The idea was that inline structs can be used like wrapped errors as well, since one may want to pass additional fields up with their error. But I think I'll put that on the back-burner for possible future improvement.

> Anyhow I think the error declaration could be cleaner and more concise if a backtrace field or additional fields are included. If more error variants are used or additional dynamic data is desired, it may look a bit bloated.
> 
> One option could be to wrap the error in a generated struct and "hide" these fields from the user. The struct itself contains the relevant metadata and error type.

I agree an additional `backtrace` field might make things bloated. I like the idea of idea of creating a wrapped typed.

```rust
error_set! {
    Error = {
        // Wraps `std::io::Error` in a type that captures backtrace. Can later but used to fulfill the `Error` trait `provide` method.
        #[trace]
        Variant1(std::io::Error),
        // Used to fulfill the `Error` trait `provide` method with a field on the wrapped error (`err`) called `backtrace`.
        #[trace = err.backtrace] 
        Variant2(std::fmt::Error),
    };
}
```

Also add `nightly` feature flag to support using the [provide](https://doc.rust-lang.org/core/error/trait.Error.html#method.provide) method





> If backtrace is no option (performance issues), the location feature could rely on [track_caller](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller). The error record with its location could be inserted into a list or just logged with error_set its tracing feature within the relevant functions or traits on error creation or conversion.

I'd rather use backtrace over location. My rational is, if one don't care about the performance hit of getting location information, one probably would want more information. And in my experience the first location is never really enough to debug an issue.

We couldn't really log or insert on creation because then we'd have to limit how users create errors. Since they can currently do things like
```rust
let x = AuthError::UserDoesNotExist {
            name: "john".to_string(),
            role: 30,
};
```
And on `into` would be too inconsistent.

We could add a new feature flag (maybe `backtrace`) that adds backtraces to all logs. e.g. something like this for the current `error` logging method

(pseudo code)
```rust
fn error(self, context: impl Display) -> Result<T, E> {
        if self.is_err() {
            #[cfg(all(feature = "tracing", not(feature="backtrace")))]
            tracing::error!("{}", context);
            #[cfg(all(feature = "tracing", feature="backtrace"))]
            tracing::error!("{}\n{}", context, std::backtrace::Backtrace::capture());
            ...
        }
        self
    }
```
The nice thing about `std::backtrace::Backtrace::capture` is it can be disabled from an env var.

or we could add a whole new set of `with_*` methods that include the backtrace, but I feel like this should be a passive things users think about, not every time they use a logging api.

**Comment by jpramosi** [comment]
Yes, it is a bit difficult to implement it without restricting the error creation. Maybe a macro/function could avoid a bit boilerplate code here.
Moreover I agree with your idea with an optional feature flag to also capture the backtrace. Together with the `log` and `tracing` feature it is a great deal in my humble opinion for stable rust.

In the meanwhile I have finished my error library! We have talked about how an alternative structure with the error location could look like. And here I can finally present you a better sketch of the idea: https://github.com/jpramosi/errore
However the design have changed a lot and it compiles only on rust nightly.
The important hot sauce is in the [trace](https://github.com/jpramosi/errore/blob/master/src/trace.rs) and [span](https://github.com/jpramosi/errore/blob/master/src/span.rs) file.
Please feel free to copy or adapt what you need, after all I got the idea here.

**Comment by mcmah309** [comment]
> In the meanwhile I have finished my error library! We have talked about how an alternative structure with the error location could look like. And here I can finally present you a better sketch of the idea: https://github.com/jpramosi/errore

Interesting implementation! Thanks for sharing. I like the idea you presented earlier and refined in your crate about logging on creation. From a glance, it looks like you use the log/location api to track every time an `Ec` is created and `into`/`?` is invoked, then collecting that information in the current `Ec`. Am I correct / missing anything else?

**Comment by jpramosi** [comment]
Yes, broadly speaking that's correct. The log macro stuff sprinkled all over the code is only there to debug errore itself. An user can append his own error subscriber to track and log errors like [here](https://github.com/jpramosi/errore/blob/master/examples/actix/src/subscriber.rs).
At first I wanted to create a similar style like the `std::io::Error` and `std::io::ErrorKind` combination for the error label, but the experimental error provide api did not allowed to mutate the error instance itself. Only with a mutex object it would be possible. But I don't want that, so I have just used the keyword `Ec` (error-context) instead of `Error` here.

**Comment by mcmah309** [comment]
I haven't decided on an api for this crate. So I'll probably be posting here for future reference with new refinements/ideas.

- Add `backtrace` flag that needs nightly. In which, `trait Error`'s `provide()` method is delegating to source by **default**.
    - Add `#[backtrace = ..]`. Overrides the above default behavior to the field on the path.
    - Add `Bt` struct that captures backtrace on creation, `provide()` method uses this backtrace, and delegates `source()` to the inner type.
- Support `source` field on inline structs.
    - Delegate `source()` method to this field. 
    - Allow inline structs that have `source` to be converted to as valid conversion targets (subsets).
- If `backtrace` flag is not enabled then backtrace annotations are ignored

```rust
error_set! {
    Error = {
        Variant1 {
            other_field: String,
            source: std::fmt::Error // delegate `source()` method to this field. On `backtrace` feature flag, delegate `provide()` as well.
        },
        #[backtrace =  type_with_backtrace.backtrace] // `provide()` is delegated to this path. If `=..` is not provided then a `backtrace` field is generated.
        Variant2 {
            other_field: String,
            source: std::io::Error,
            type_with_backtrace: Type,
        },
        #[backtrace] // Essentially creates a `Variant3(Bt(std::io::Error)),`
        Variant3(std::io::Error),
    };
}
```

**Comment by mcmah309** [comment]
Another alternative to `source: ..` field is a `#[source(..)]` annotation on inline structs, for generating the field. This opens up the possibility of `#[source = ..]` for sources not named "source" and is more inline with the `#[backtrace]` annotation since that may generate a field. It also makes the relationship to `source` more explicit.

**Comment by mcmah309** [comment]
The first step related to this this has been released (source structs) -
https://github.com/mcmah309/error_set/releases/tag/v0.7.0
